### PR TITLE
init-sync: refactor block fetching

### PR DESCRIPTION
**Pull request by farazdagi**
Part of #4815

---

# Description

Reference implementation of non-blocking initial synchronization, where producer-consumer queue is used to separate block fetching from processing (and allow them to run concurrently).  More details are outlined in the [design document](https://docs.google.com/document/d/1mm__vLgJfkq25iryIAqWNSgFNfaXqU2tM7P31e8jyf8/edit?usp=sharing).

This PR, handles only fetching part of the requirements:
- round robin initial sync **has been patched** to use new fetcher
- while new fetcher can be run in a non-blocking mode, since we've decided to move in incremental steps, almost one-to-one functionality port was made. So, ideally, everything should be run as before, but the code underneath is decoupled.

### CODE REVIEW

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+	"github.com/prysmaticlabs/prysm/beacon-chain/flags"
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+	prysmsync "github.com/prysmaticlabs/prysm/beacon-chain/sync"
+	p2ppb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/mathutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/sirupsen/logrus"
+)
+
+// blocksFetcherConfig is a config to setup block fetcher
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]:
```suggestion
// blocksFetcherConfig is a config to setup the block fetcher.
```
Make sure comments are always full sentences with periods 😄

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+		quit:        make(chan struct{}),
+	}
+}
+
+// start boots up the fetcher, which starts listening for incoming fetch requests.
+func (f *blocksFetcher) start() {
+	go f.loop()
+}
+
+// stop terminates all fetcher operations
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]:
```suggestion
// stop terminates all fetcher operations.
```

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+	chain       blockchainService
+	p2p         p2p.P2P
+	rateLimiter *leakybucket.Collector
+
+	requests chan *fetchRequestParams // incoming fetch requests
+	out      chan *fetchRequestResult // outgoing responses
+	quit     chan struct{}            // termination notifier
+}
+
+// fetchRequestParams holds parameters necessary to schedule a fetch request
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]:
```suggestion
// fetchRequestParams holds parameters necessary to schedule a fetch request.
```

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+	defer close(f.out)
+
+	randGenerator := rand.New(rand.NewSource(time.Now().Unix()))
+	highestFinalizedSlot := helpers.StartSlot(f.highestFinalizedEpoch() + 1)
+	maxPeersToSync := params.BeaconConfig().MaxPeersToSync
+
+	for {
+		select {
+		case <-f.ctx.Done():
+			// upstream context is done
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]:
```suggestion
			// Upstream context is done.
```

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+			return
+		case <-f.quit:
+			// terminating abort all operations
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]:
```suggestion
			// Terminating abort all operations.
```

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+				err := errors.Errorf("requested a start slot of %d which is greater than the next highest slot of %d", req.start, highestFinalizedSlot)
+				log.WithError(err).Debug("Block fetch request failed")
+				f.out <- &fetchRequestResult{
+					params: req,
+					err:    err,
+				}
+				continue
+			}
+
+			// shuffle peers to prevent a bad peer from
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]:
```suggestion
			// Shuffle peers to prevent a bad peer from
```

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+			// stalling sync with invalid blocks
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]:
```suggestion
			// stalling sync with invalid blocks.
```

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+
+// blocksFetcher is a service to fetch chain data from peers.
+// On an incoming requests, requested block range is evenly divided
+// among available peers (for fair network load distribution).
+type blocksFetcher struct {
+	ctx         context.Context
+	chain       blockchainService
+	p2p         p2p.P2P
+	rateLimiter *leakybucket.Collector
+
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]:
```suggestion
```

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+	p2pRequests.Add(len(peers))
+	go func() {
+		p2pRequests.Wait()
+		close(blocksChan)
+	}()
+
+	// Short circuit start far exceeding the highest finalized epoch in some infinite loop.
+	highestFinalizedSlot := helpers.StartSlot(finalizedEpoch + 1)
+	if start > highestFinalizedSlot {
+		return nil, errors.Errorf("attempted to ask for a start slot of %d which is greater than the next highest slot of %d", start, highestFinalizedSlot)
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]:
```suggestion
		return nil, errors.Errorf("did not expect request for start slot of %d which is greater than expected %d", start, highestFinalizedSlot)
```

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+	}
+
+	count = mathutil.Min(count, helpers.StartSlot(finalizedEpoch+1)-start+1) // do not overflow the finalized epoch
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]: Can you move this comment to the line above?

File: beacon-chain/sync/initial-sync/blocks_fetcher_test.go [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+	mock "github.com/prysmaticlabs/prysm/beacon-chain/blockchain/testing"
+	dbtest "github.com/prysmaticlabs/prysm/beacon-chain/db/testing"
+	p2pt "github.com/prysmaticlabs/prysm/beacon-chain/p2p/testing"
+	stateTrie "github.com/prysmaticlabs/prysm/beacon-chain/state"
+	p2ppb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/sliceutil"
+	"github.com/sirupsen/logrus"
+)
+
+func TestBlocksFetcher(t *testing.T) {
```
Comment (rauljordan) [inline beacon-chain/sync/initial-sync/blocks_fetcher_test.go]: One thing that could be very helpful for tests like this that are testing behavior of an entire service is to assert for logs. You can see an example here: https://github.com/prysmaticlabs/prysm/blob/b647ca5dd2f775742f1dc146ecfc58bed2aa2bd8/slasher/beaconclient/historical_data_retrieval_test.go#L78 we basically check easily if a log we want has occurred. You can use that to test edge cases or uncovered lines of your logic.

File: beacon-chain/sync/initial-sync/blocks_fetcher_test.go [hunk 12]
```diff
@@ -0,0 +1,362 @@
+package initialsync
```
Comment (rauljordan) [inline beacon-chain/sync/initial-sync/blocks_fetcher_test.go]: One suggestion I have is to add more granular tests if you end up splitting up a lot of the large functions in block_fetcher.go. This would give you greater control of ensuring certain things fail as expected, especially the concurrent functions. It's always nice indeed to have one big function that tests the full behavior of a service runtime, so no need to delete it, but would be nice to have smaller tests as well.

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+
+// newBlocksFetcher creates ready to use fetcher
+func newBlocksFetcher(cfg *blocksFetcherConfig) *blocksFetcher {
+	return &blocksFetcher{
+		ctx:         cfg.ctx,
+		chain:       cfg.chain,
+		p2p:         cfg.p2p,
+		rateLimiter: cfg.rateLimiter,
+		requests:    make(chan *fetchRequestParams),
+		out:         make(chan *fetchRequestResult),
```
[the remark this answers is not in the document]
Comment (farazdagi) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go]: I keep everything unbuffered, to make sure everything works in a blocking (full-buffer) mode, w/o deadlocks. Once we will have all the moving parts in place (+ queue/processor), not only will we make those buffered, we will expose buffer size as package variable, allowing us to tune things in production.

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+	p2ppb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/mathutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/sirupsen/logrus"
+)
+
+// blocksFetcherConfig is a config to setup block fetcher
+type blocksFetcherConfig struct {
+	ctx         context.Context
+	chain       blockchainService
```
[the remark this answers is not in the document]
Comment (farazdagi) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go]: Thank you, great suggestion. Replaced.

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+	randGenerator := rand.New(rand.NewSource(time.Now().Unix()))
+	highestFinalizedSlot := helpers.StartSlot(f.highestFinalizedEpoch() + 1)
+	maxPeersToSync := params.BeaconConfig().MaxPeersToSync
+
+	for {
+		select {
+		case <-f.ctx.Done():
+			// upstream context is done
+			return
+		case <-f.quit:
```
[the remark this answers is not in the document]
Comment (farazdagi) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go]: Well, no - at least it is not as I've intended. Closed `f.quit` is a normal condition, where fetcher was just stopped via `fetcher.stop()`, which is not an error condition by itself.

File: beacon-chain/sync/initial-sync/blocks_fetcher_test.go [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+				fetcher.stop()
+			}()
+
+			processFetchedBlocks := func() ([]*eth.SignedBeaconBlock, error) {
+				defer cancel()
+				var unionRespBlocks []*eth.SignedBeaconBlock
+
+				for {
+					select {
+					case <-time.After(5 * time.Second): // TODO(4815): Temporary safeguard, remove
```
[the remark this answers is not in the document]
Comment (farazdagi) [reply beacon-chain/sync/initial-sync/blocks_fetcher_test.go]: Agreed! The only reason it is structured like that is that, I wanted to make it as close to round robin tests as possible, as I planned to swap the current implementation of round robing with my refactored code - so wanted to increase chances it will work w/o issues. Still, we can refactor further, and split the code into more manageable chunks (for both using and tests).

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 17]
```diff
@@ -78,123 +86,37 @@ func (s *Service) roundRobinSync(genesis time.Time) error {
...
 			log.WithField("finalizedEpoch", finalizedEpoch).Debug("Requested block range is greater than the finalized epoch")
 			break
 		}
+		startBlock += skippedBlocks
+
+		// Operations below are simple port of the existing functionality, where synchronization is assumed.
+		// This will be refactored and implemented w/i queue, once that component is ready.
+		request := func() ([]*eth.SignedBeaconBlock, error) {
+			select {
+			case <-time.After(60 * time.Second): // extra pre-caution
```
[the remark this answers is not in the document]
Comment (nisdas) [reply beacon-chain/sync/initial-sync/round_robin.go]: +1

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 18]
```diff
@@ -78,123 +86,37 @@ func (s *Service) roundRobinSync(genesis time.Time) error {
...
+				return nil, errors.New("it takes too long to receive data from fetcher, unblock")
+			case resp, ok := <-fetcher.iter(): // fetcher will stop when upstream ctx is closed
+				if !ok { // channel closed
```
[the remark this answers is not in the document]
Comment (nisdas) [reply beacon-chain/sync/initial-sync/round_robin.go]: +1 , maybe a check for upstream context would be more appropriate here.

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+			return
+		case req := <-f.requests:
+			root, finalizedEpoch, peers := f.p2p.Peers().BestFinalized(maxPeersToSync, helpers.SlotToEpoch(f.headFetcher.HeadSlot()))
+			log.WithFields(logrus.Fields{
+				"request":        req,
+				"finalizedEpoch": finalizedEpoch,
+				"numPeers":       len(peers),
+			}).Debug("Block fetcher receives request")
+
+			if len(peers) == 0 {
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]: I think this piece of code still exists in the `round_robin` main function. Either have it here or there.

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+				f.receivedFetchResponses <- &fetchRequestResponse{
+					params: req,
+					err:    err,
+				}
+				continue
+			}
+
+			// Shuffle peers to prevent a bad peer from
+			// stalling sync with invalid blocks.
+			randGenerator.Shuffle(len(peers), func(i, j int) {
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]: same here we are shuffling our peer set again over here.

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+
+			// shuffle peers to prevent a bad peer from
+			// stalling sync with invalid blocks
+			randGenerator.Shuffle(len(peers), func(i, j int) {
+				peers[i], peers[j] = peers[j], peers[i]
+			})
+
+			// TODO(4815): Consider splitting peers into sets (when there are many of them),
+			// so that block fetching is for lesser chunks and is less affected by some slow peer
+			go func(root []byte, finalizedEpoch, start, count uint64, peers []peer.ID) {
```
[the remark this answers is not in the document]
Comment (nisdas) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go]: I dont think this needs to be an anonymous function, or why not just do 
```go
go func(){
				resp, err := f.processFetchRequest(root, finalizedEpoch, req.start, req.count, peers)
				if err != nil {
					log.WithError(err).Debug("Block fetch request failed")
					f.receivedFetchResponses <- &fetchRequestResponse{
						params: req,
						err:    err,
					}
					return
				}

				f.receivedFetchResponses <- &fetchRequestResponse{
					params: req,
					blocks: resp,
				}
}()
```
if you want to make it non-blocking

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+			}(root, finalizedEpoch, req.start, req.count, peers)
+		}
+	}
+}
+
+// scheduleRequest adds request to incoming queue.
+// Should be non-blocking, as actual request processing is done asynchronously.
+func (f *blocksFetcher) scheduleRequest(req *fetchRequestParams) {
+	go func() { // non-blocking, we can re-throw requests within consuming method
+		select {
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]: do we actually need a select here ? it seems a bit redundant . we can check if quit is a closed channel at the start and if not pass the request to our `requests` channel. Since that is non-buffered maybe enclose it in a separate go-routine.

File: beacon-chain/sync/initial-sync/blocks_fetcher.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go]: I think you need to have arguments since this is in a loop

[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go]: This does look like a good candidate for refactoring though.

Please consider moving to a new method as @rauljordan suggested and adding tests for this method's functionality.

[the remark this answers is not in the document]
Comment (farazdagi) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go]: Extracting this method atm :) Certainly makes things a) more clear b) easier to test.

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/sync/initial-sync/round_robin.go]: Can you use a context deadline instead?
